### PR TITLE
Refactor tests PR

### PR DESCRIPTION
**Pull request by calebcartwright**
Start of major refactoring that includes both tests and app code to create isolated functions for handling the specifics of respective project types. 

Note this PR also includes the minor fix to #78, and that right now I have commented out the original test suite from the core index-tests file as I believe we can potentially extract those as component-integration tests.

The big discussion topic is how we should go about unit testing these functions. What I've done so far in the [VSTS Task Tests](https://github.com/swellaby/generator-swell/compare/refactor-tests?expand=1#diff-32a8f5ec5aeece25e5f4b37f82670ed5) is essentially just a direct port. However, in a typical (non yeoman) world I don't think these would classify as unit tests, given that the setup for this is still using the yeoman helpers to do testing from the top down (through yeomen & index). With the new breakdown I think it would be possible to go super-stubs/super-mocks to do a 'true' unit test

@beverts312  Do you think this is sufficient in the yeoman world, or should we go the super-stubs route? I'm inclined to tackle the super route..

### CODE REVIEW

**Comment by beverts312** [comment]
@calebcartwright 
I really like the increased readability in the generator code.  
Are we going to keep the helper functions (not related to a generator i.e `_validateDirectoryName`) in `index.ts` ?  

I do think tests of this nature are sufficient for the yeoman world in terms of actually getting the job done. That being said, I do think the super-stubs route is proper way to do it (and would be pretty sexy).

**Comment by calebcartwright** [comment]
I've already gone down the path of going for "pure" unit tests. The setup is rather painful (which I assume is why yeoman has multiple separate libraries to facilitate "unit" testing) but I think I can encapsulate all that pain in one place. 

For the first pass I am thinking we extract all the project specific stuff, and leave the rest in index. Before merging this we can take a look at what is left in index and see if it makes sense to extract as well.

**Comment by calebcartwright** [comment]
Most of these commits are more general repo items, not necessarily about the broad umbrella of this issue/PR. I know where I want to take things now so shouldn't be too much longer before there will be some more interesting things to look at.

**Comment by calebcartwright** [comment]
@beverts312 the VSTS portion of the refactor (including tests) is just about done. A few things to note/discuss on this commit

- I uncommented out the original VSTS Task test suite in index-tests.ts. I haven't written all the "true" unit tests for the main index.ts file yet, so I needed those originals to still hit the 100% coverage threshold. 
- I ported that same suite of tests, which is really a suite of e2e component integration tests from index -> vsts -> yo, into a new location in the [component integration tests directory ](https://github.com/swellaby/generator-swell/blob/eb9284e5546c33b3e3a50e4ba5adf8e6494310b0/test/component-integration/e2e/index-vsts-tests.ts)
- all of the nasty yo stubs, shared content, etc. necessary for the new, true unit tests are encapsulated  [here](https://github.com/swellaby/generator-swell/blob/eb9284e5546c33b3e3a50e4ba5adf8e6494310b0/test/test-helpers.ts)
- the new vsts unit tests (100% coverage) are [here](https://github.com/swellaby/generator-swell/blob/eb9284e5546c33b3e3a50e4ba5adf8e6494310b0/test/unit/app/vsts-tests.ts), although note that the last two tests w/valid inputs have super simple assertions that I will flesh out with real ones
- the far more succinct VSTS related content is [here](https://github.com/swellaby/generator-swell/blob/refactor-tests/generators/app/vsts.ts)

**Comment by beverts312** [comment]
This looks good. 
Do we need the `unit/app` & `component-integration/e2e` directories? Or could we drop the `app` & `e2e` directories and move the inner files up a level?  

Not a big deal but I am having trouble coming up with subfolders would need to exist at that level in the folder structure.

**Comment by calebcartwright** [comment]
I spent a while thinking about the original test/app structure, and already felt a twinge of guilt about excluding the generator directory as I tend to buy into the approach of having tests (unit tests anyway) matching the file name & directory structure of the application code they are testing. If we were ever to add any additional sub-generator cabapilties to this there would be an additional set of directories under `generators` at the same level as `app`  (for example `generators\app` and `generators\express-controller`). while I don't anticipate doing this anytime soon (or perhaps ever) there are certainly many value add scenarios I could envision, especially with the growing number of project types the core generator supports. 

i'm definitely open to changes on component-integration. my thinking was that if we were so inclined there a lot of other component tests that could be written for the generator codebase. the current tests we have IMO really do go top to bottom through multiple components, but we could easily have additional tests that validate the integration between just index & vsts with the yo components stubbed out.

**Comment by calebcartwright** [comment]
@beverts312 - getting pretty close here. Both the [vsts tests](https://github.com/swellaby/generator-swell/blob/6b699d7b77553874096a91a31bdcd17a818f6056/test/unit/app/vsts-tests.ts) and the [vs code tests](https://github.com/swellaby/generator-swell/blob/6b699d7b77553874096a91a31bdcd17a818f6056/test/unit/app/vscode-tests.ts) are done and should be good examples of what the "true" unit tests will look like. 

All of the refactored app code for the individual project types being maintained in their own respective files is done, but as before need to add a few more validations and/or tests on them to finish testing all paths (similar to the aforementioned tests for vsts and vs code) 

The swellaby-generator file/class is all set, but commented out as I haven't written the tests yet as I have been intentionally procrastinating. Next big step is to ensure the refactor we discussed is possible, and then flipping over to the new model.

**Comment by calebcartwright** [comment]
@beverts312 and anyone else following this.. core refactoring is done. The only thing left is pulling out our old "unit" tests into dedicated suites under component integration (which is what they actually are). this branch is so out of date though that I'm recommending to go ahead and merge (which I will do after resolving conflicts with master). and then I will extend the scope of #83 to cover the component integration tests et al.

Concerns?

File: generators/app/index-tbr.ts [hunk 1]
```diff
@@ -0,0 +1,272 @@
+// /* tslint:disable:function-name */
```
Comment (calebcartwright) [inline generators/app/index-tbr.ts]: will remove this file as part of #83

**Pull request merged**